src/pe/certificate_table.rs

```diff
@@ -0,0 +1,90 @@
+    type Error = error::Error;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        Ok(match value {
+            x if x == CertificateAttributeRevision::Revision1_0 as u16 => CertificateAttributeRevision::Revision1_0,
+            x if x == CertificateAttributeRevision::Revision2_0 as u16 => CertificateAttributeRevision::Revision2_0,
+            _ => return Err(error::Error::Malformed("Invalid certificate attribute revision".to_string()))
+        })
+    }
+}
+
+#[repr(u16)]
+#[derive(Debug)]
+enum CertificateAttributeType {
+    X509 = 0x0001,
+    PkcsSignedData = 0x0002,
+    Reserved1 = 0x0003,
+    TsStackSigned = 0x0004,
+}
+
+impl TryFrom<u16> for CertificateAttributeType {
+    type Error = error::Error;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        Ok(match value {
+            x if x == CertificateAttributeType::X509 as u16 => CertificateAttributeType::X509,
+            x if x == CertificateAttributeType::PkcsSignedData as u16 => CertificateAttributeType::PkcsSignedData,
+            x if x == CertificateAttributeType::Reserved1 as u16 => CertificateAttributeType::Reserved1,
+            x if x == CertificateAttributeType::TsStackSigned as u16 => CertificateAttributeType::TsStackSigned,
+            _ => return Err(error::Error::Malformed("Invalid certificate attribute type".to_string()))
+        })
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Pread)]
+pub struct CertificateAttributeHeader {
+    length: u32,
+    revision: u16,
+    certificate_type: u16
+}
+
+pub const CERTIFICATE_DATA_OFFSET: u32 = 8;
+#[derive(Debug)]
+pub struct CertificateAttribute<'a> {
+    length: u32,
+    revision: CertificateAttributeRevision,
+    certificate_type: CertificateAttributeType,
+    certificate: &'a [u8]
+}
+
+impl<'a> CertificateAttribute<'a> {
+    pub fn parse(bytes: &'a [u8], current_offset: &mut usize) -> Result<CertificateAttribute<'a>, error::Error> {
+        let header: CertificateAttributeHeader = bytes.gread_with(current_offset, scroll::LE)?;
+        let mut cert_size = (header.length - CERTIFICATE_DATA_OFFSET) as usize;
+        // Round to the next 8-bytes.
+        cert_size = (cert_size + 7) & !7;
+        let attr = Self {
+            length: header.length,
+            revision: header.revision.try_into()?,
+            certificate_type: header.certificate_type.try_into()?,
+            certificate: &bytes[*current_offset..(*current_offset + cert_size)]
+        };
+        *current_offset += cert_size as usize;
+        Ok(attr)
+    }
+}
+
+pub type CertificateDirectoryTable<'a> = Vec<CertificateAttribute<'a>>;
+pub fn enumerate_certificates<'a>(bytes: &'a [u8], table_virtual_address: u32, table_size: u32) -> Result<CertificateDirectoryTable<'a>, error::Error> {
+    let mut current_offset = table_virtual_address as usize;
+    let mut attrs = vec![];
+
+    while (current_offset - table_virtual_address as usize) != (table_size as usize) {
```

Comment:
can this overflow? (e.g., what if table_virtual_address is > current_offset); also this has the potential for an infinite loop, I would prefer this to be rewritten so that it is not possible (goblin routinely gets passed binaries with bad values, incorrect sizes, fuzzed, etc., so a while loop requiring exact match of a size could have many inputs that lead to infinite loop)

Reply:
Technically, it can, because `as` conversions are really super unsafe.
I transformed the whole code into something that uses saturation semantics (i.e. staying at MAX rather than wrapping back) which, I think, is the semantic we want.

By virtue of various checks, infinite loop and too many manipulations should not be possible through an adversarial binary. (this needs to be tested though, honestly)

Also, it depends a lot on the actual size of `usize`, if it's less than 32 bits, annoying things could arise.